File: server/storage.js
// ===== Object storage (S3) — MỘT client duy nhất cho mọi môi trường =====
// local dev -> MinIO;  staging/UAT/prod -> Supabase Storage / AWS S3.
// Chỉ khác endpoint + credentials qua ENV; KHÔNG rẽ nhánh framework theo môi trường.
const { S3Client, PutObjectCommand, GetObjectCommand, DeleteObjectCommand } = require('@aws-sdk/client-s3');
const { getSignedUrl } = require('@aws-sdk/s3-request-presigner');

// Fail-fast: bắt buộc cấu hình S3 ở MỌI môi trường (không có fallback base64/đĩa)
const NEED = ['S3_ENDPOINT', 'S3_REGION', 'S3_ACCESS_KEY', 'S3_SECRET_KEY', 'S3_CCCD_BUCKET', 'S3_INTRO_BUCKET'];
const missing = NEED.filter(k => !process.env[k]);
if (missing.length) {
  throw new Error('Thiếu cấu hình object storage (S3): ' + missing.join(', ') +
    '. Local: chạy "docker compose up -d" (MinIO) rồi điền S3_* trong .env.');
}

const CCCD_BUCKET = process.env.S3_CCCD_BUCKET;
const INTRO_BUCKET = process.env.S3_INTRO_BUCKET;

const s3 = new S3Client({
  endpoint: process.env.S3_ENDPOINT,
  region: process.env.S3_REGION,
  credentials: { accessKeyId: process.env.S3_ACCESS_KEY, secretAccessKey: process.env.S3_SECRET_KEY },
  forcePathStyle: true, // MinIO & Supabase Storage đều dùng path-style
});

// CHỈ chấp nhận ảnh raster an toàn. KHÔNG nhận SVG (có thể chứa <script> -> XSS khi proxy).
const EXT = { 'image/jpeg': 'jpg', 'image/png': 'png', 'image/webp': 'webp', 'image/gif': 'gif' };

function parseDataUrl(dataUrl) {
  const m = /^data:(image\/[\w.+-]+);base64,(.+)$/s.exec(dataUrl || '');
  if (!m) return null;
  const ext = EXT[m[1]];
  if (!ext) return null; // loại type ngoài whitelist (vd svg) -> caller sẽ báo "Ảnh không hợp lệ"
  return { contentType: m[1], ext, buffer: Buffer.from(m[2], 'base64') };
}

async function putBuffer(bucket, key, buffer, contentType) {
  await s3.send(new PutObjectCommand({
    Bucket: bucket, Key: key, Body: buffer, ContentType: contentType, CacheControl: 'public, max-age=3600',
  }));
  return key;
}

async function putDataUrl(bucket, key, dataUrl) {
  const p = parseDataUrl(dataUrl);
  if (!p) { const e = new Error('Ảnh không hợp lệ (chỉ nhận JPG/PNG/WEBP/GIF)'); e.status = 400; throw e; }
  return putBuffer(bucket, key, p.buffer, p.contentType);
}

// Tài liệu PDF (nội quy ký túc xá). Tách riêng khỏi parseDataUrl để KHÔNG vô tình cho phép
// upload PDF vào những chỗ chỉ được nhận ảnh (ảnh CCCD, ảnh trang giới thiệu).
function parsePdfDataUrl(dataUrl) {
  const m = /^data:application\/pdf;base64,(.+)$/s.exec(dataUrl || '');
  if (!m) return null;
  const buffer = Buffer.from(m[1], 'base64');
  // Kiểm chữ ký file thật, không tin mỗi cái nhãn data: — đổi nhãn là qua mặt được.
  if (buffer.slice(0, 5).toString('latin1') !== '%PDF-') return null;
  return { contentType: 'application/pdf', ext: 'pdf', buffer };
}
async function putPdfDataUrl(bucket, key, dataUrl) {
  const p = parsePdfDataUrl(dataUrl);
  if (!p) { const e = new Error('Tệp không hợp lệ — chỉ nhận file PDF'); e.status = 400; throw e; }
  return putBuffer(bucket, key, p.buffer, p.contentType);
}

// Lấy object để stream ra client (proxy). Trả stream + metadata; ném lỗi nếu không có.
async function getObject(bucket, key) {
  const r = await s3.send(new GetObjectCommand({ Bucket: bucket, Key: key }));
  return { body: r.Body, contentType: r.ContentType, contentLength: r.ContentLength, etag: r.ETag };
}

async function deleteObject(bucket, key) {
  if (!key) return;
  await s3.send(new DeleteObjectCommand({ Bucket: bucket, Key: key }));
}

// Link ký hạn (không dùng cho luồng hiển thị mặc định — app proxy — nhưng để sẵn tiện ích)
async function signedGetUrl(bucket, key, expiresIn = 3600) {
  return getSignedUrl(s3, new GetObjectCommand({ Bucket: bucket, Key: key }), { expiresIn });
}

module.exports = {
  s3, CCCD_BUCKET, INTRO_BUCKET,
  parseDataUrl, parsePdfDataUrl, putBuffer, putDataUrl, putPdfDataUrl, getObject, deleteObject, signedGetUrl,
};

File: internal/storage/storage.go
// Package storage — object storage (S3) một client cho mọi môi trường (MinIO/Supabase/AWS).
// Port từ server/storage.js. Chỉ khác endpoint+credentials qua ENV.
package storage

import (
	"bytes"
	"context"
	"encoding/base64"
	"fmt"
	"io"
	"regexp"
	"time"

	"github.com/aws/aws-sdk-go-v2/aws"
	awsconfig "github.com/aws/aws-sdk-go-v2/config"
	"github.com/aws/aws-sdk-go-v2/credentials"
	"github.com/aws/aws-sdk-go-v2/service/s3"
	appconfig "ktx/internal/config"
)

type Storage struct {
	client      *s3.Client
	presign     *s3.PresignClient
	CccdBucket  string
	IntroBucket string
}

// New dựng client S3. Trả lỗi nếu thiếu cấu hình (server/storage.js:8-13).
func New(ctx context.Context, cfg *appconfig.Config) (*Storage, error) {
	missing := []string{}
	for k, v := range map[string]string{
		"S3_ENDPOINT": cfg.S3Endpoint, "S3_REGION": cfg.S3Region, "S3_ACCESS_KEY": cfg.S3AccessKey,
		"S3_SECRET_KEY": cfg.S3SecretKey, "S3_CCCD_BUCKET": cfg.S3CccdBucket, "S3_INTRO_BUCKET": cfg.S3IntroBucket,
	} {
		if v == "" {
			missing = append(missing, k)
		}
	}
	if len(missing) > 0 {
		return nil, fmt.Errorf("thiếu cấu hình object storage (S3): %v", missing)
	}
	awsCfg, err := awsconfig.LoadDefaultConfig(ctx,
		awsconfig.WithRegion(cfg.S3Region),
		awsconfig.WithCredentialsProvider(credentials.NewStaticCredentialsProvider(cfg.S3AccessKey, cfg.S3SecretKey, "")),
	)
	if err != nil {
		return nil, err
	}
	client := s3.NewFromConfig(awsCfg, func(o *s3.Options) {
		o.BaseEndpoint = aws.String(cfg.S3Endpoint)
		o.UsePathStyle = true // MinIO & Supabase Storage dùng path-style
	})
	return &Storage{client: client, presign: s3.NewPresignClient(client), CccdBucket: cfg.S3CccdBucket, IntroBucket: cfg.S3IntroBucket}, nil
}

// EXT + magic bytes: chỉ nhận ảnh raster an toàn (KHÔNG SVG). server/storage.js:26-41
var extByType = map[string]string{"image/jpeg": "jpg", "image/png": "png", "image/webp": "webp", "image/gif": "gif"}

func khopMagic(ext string, b []byte) bool {
	if len(b) < 4 {
		return false
	}
	switch ext {
	case "jpg":
		return b[0] == 0xFF && b[1] == 0xD8 && b[2] == 0xFF
	case "png":
		return b[0] == 0x89 && b[1] == 0x50 && b[2] == 0x4E && b[3] == 0x47
	case "gif":
		return string(b[0:3]) == "GIF"
	case "webp":
		return len(b) >= 12 && string(b[0:4]) == "RIFF" && string(b[8:12]) == "WEBP"
	}
	return false
}

var reImageDataURL = regexp.MustCompile(`(?s)^data:(image/[\w.+-]+);base64,(.+)$`)
var rePdfDataURL = regexp.MustCompile(`(?s)^data:application/pdf;base64,(.+)$`)

type Parsed struct {
	ContentType string
	Ext         string
	Buffer      []byte
}

// ParseDataUrl: data:image/...;base64 -> buffer, kiểm magic bytes. server/storage.js:43-51
func ParseDataUrl(dataURL string) *Parsed {
	m := reImageDataURL.FindStringSubmatch(dataURL)
	if m == nil {
		return nil
	}
	ext, ok := extByType[m[1]]
	if !ok {
		return nil
	}
	buf, err := base64.StdEncoding.DecodeString(m[2])
	if err != nil || !khopMagic(ext, buf) {
		return nil
	}
	return &Parsed{ContentType: m[1], Ext: ext, Buffer: buf}
}

// ParsePdfDataUrl: data:application/pdf;base64 -> buffer, kiểm %PDF-. server/storage.js:68-75
func ParsePdfDataUrl(dataURL string) *Parsed {
	m := rePdfDataURL.FindStringSubmatch(dataURL)
	if m == nil {
		return nil
	}
	buf, err := base64.StdEncoding.DecodeString(m[1])
	if err != nil || len(buf) < 5 || string(buf[0:5]) != "%PDF-" {
		return nil
	}
	return &Parsed{ContentType: "application/pdf", Ext: "pdf", Buffer: buf}
}

// HTTPError: lỗi có status cho handler.
type HTTPError struct {
	Status int
	Msg    string
}

func (e *HTTPError) Error() string { return e.Msg }

func (st *Storage) PutBuffer(ctx context.Context, bucket, key string, buffer []byte, contentType string) (string, error) {
	cc := "public, max-age=3600"
	_, err := st.client.PutObject(ctx, &s3.PutObjectInput{
		Bucket: &bucket, Key: &key, Body: bytes.NewReader(buffer), ContentType: &contentType, CacheControl: &cc,
	})
	if err != nil {
		return "", err
	}
	return key, nil
}

// PutDataUrl: server/storage.js:60-64
func (st *Storage) PutDataUrl(ctx context.Context, bucket, key, dataURL string) (string, error) {
	p := ParseDataUrl(dataURL)
	if p == nil {
		return "", &HTTPError{Status: 400, Msg: "Ảnh không hợp lệ (chỉ nhận JPG/PNG/WEBP/GIF)"}
	}
	return st.PutBuffer(ctx, bucket, key, p.Buffer, p.ContentType)
}

// PutPdfDataUrl: server/storage.js:76-80
func (st *Storage) PutPdfDataUrl(ctx context.Context, bucket, key, dataURL string) (string, error) {
	p := ParsePdfDataUrl(dataURL)
	if p == nil {
		return "", &HTTPError{Status: 400, Msg: "Tệp không hợp lệ — chỉ nhận file PDF"}
	}
	return st.PutBuffer(ctx, bucket, key, p.Buffer, p.ContentType)
}

type Object struct {
	Body          io.ReadCloser
	ContentType   string
	ContentLength int64
	ETag          string
}

// GetObject: lấy object để proxy stream. server/storage.js:83-86
func (st *Storage) GetObject(ctx context.Context, bucket, key string) (*Object, error) {
	r, err := st.client.GetObject(ctx, &s3.GetObjectInput{Bucket: &bucket, Key: &key})
	if err != nil {
		return nil, err
	}
	o := &Object{Body: r.Body}
	if r.ContentType != nil {
		o.ContentType = *r.ContentType
	}
	if r.ContentLength != nil {
		o.ContentLength = *r.ContentLength
	}
	if r.ETag != nil {
		o.ETag = *r.ETag
	}
	return o, nil
}

func (st *Storage) DeleteObject(ctx context.Context, bucket, key string) error {
	if key == "" {
		return nil
	}
	_, err := st.client.DeleteObject(ctx, &s3.DeleteObjectInput{Bucket: &bucket, Key: &key})
	return err
}

// SignedGetUrl: link ký hạn (tiện ích). server/storage.js:94-96
func (st *Storage) SignedGetUrl(ctx context.Context, bucket, key string, expires time.Duration) (string, error) {
	req, err := st.presign.PresignGetObject(ctx, &s3.GetObjectInput{Bucket: &bucket, Key: &key}, s3.WithPresignExpires(expires))
	if err != nil {
		return "", err
	}
	return req.URL, nil
}
